### PR TITLE
Add support for 3.15

### PR DESCRIPTION
**Pull request by martskins**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1078,6 +1096,19 @@ pub struct CompletionItemCapability {
...
     #[serde(skip_serializing_if = "Option::is_none")]
     pub preselect_support: Option<bool>,
+
+    /**
+     * Client supports the tag property on a completion item. Clients supporting
+     * tags have to handle unknown tags gracefully. Clients especially need to
+     * preserve unknown tags when sending a completion item back to the server in
+     * a resolve call.
+     */
+    pub tag_support: Option<TagSupport<CompletionItemTag>>,
```
Comment (kjeremy) [inline src/lib.rs]: `#[serde(skip_serializing_if = "Option::is_none")]`

File: src/lib.rs [hunk 2]
```diff
@@ -1181,6 +1212,14 @@ pub struct SignatureHelpCapability {
...
     #[serde(skip_serializing_if = "Option::is_none")]
     pub signature_information: Option<SignatureInformationSettings>,
+
+    /**
+     * The client supports to send additional context information for a
+     * `textDocument/signatureHelp` request. A client that opts into
+     * contextSupport will also support the `retriggerCharacters` on
+     * `SignatureHelpOptions`.
+     */
+    pub context_support: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: `#[serde(skip_serializing_if = "Option::is_none")]`

File: src/lib.rs [hunk 3]
```diff
@@ -1191,6 +1230,17 @@ pub struct PublishDiagnosticsCapability {
      */
     #[serde(skip_serializing_if = "Option::is_none")]
     pub related_information: Option<bool>,
+
+    /// Client supports the tag property to provide meta data about a diagnostic.
+    /// Clients supporting tags have to handle unknown tags gracefully.
+    pub tag_support: Option<TagSupport<DiagnosticTag>>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 4]
```diff
@@ -1369,6 +1419,17 @@ pub struct ClientCapabilities {
 pub struct InitializeResult {
     /// The capabilities the language server provides.
     pub capabilities: ServerCapabilities,
+
+    /// The capabilities the language server provides.
+    pub server_info: Option<ServerInfo>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 5]
```diff
@@ -1443,6 +1504,52 @@ pub struct SignatureHelpOptions {
...
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SignatureHelpContext {
+    ///  Action that caused signature help to be triggered.
+    pub trigger_kind: SignatureHelpTriggerKind,
+
+    /// Character that caused signature help to be triggered.
+    /// This is undefined when `triggerKind !== SignatureHelpTriggerKind.TriggerCharacter`
+    pub trigger_character: Option<String>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 6]
```diff
@@ -1570,6 +1677,9 @@ pub struct CodeActionCapability {
     /// response of the `textDocument/codeAction` request.
     #[serde(skip_serializing_if = "Option::is_none")]
     pub code_action_literal_support: Option<CodeActionLiteralSupport>,
+
+    /// Whether code action supports the `isPreferred` property.
+    pub is_preferred_support: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 7]
```diff
@@ -2133,13 +2255,21 @@ pub struct PublishDiagnosticsParams {
 
     /// An array of diagnostic information items.
     pub diagnostics: Vec<Diagnostic>,
+
+    /// Optional the version number of the document the diagnostics are published for.
+    pub version: Option<i64>,
```
Comment (kjeremy) [inline src/lib.rs]: I think there's a `Number` field or typedef somewhere in this crate that you can use.

Comment (kjeremy) [reply src/lib.rs]: And of course #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 8]
```diff
@@ -3359,8 +3521,8 @@ pub struct WorkDoneProgressBegin {
     pub title: String,
 
     /// Controls if a cancel button should show to allow the user to cancel the
-	/// long running operation. Clients that don't support cancellation are allowed
-	/// to ignore the setting.
+    /// long running operation. Clients that don't support cancellation are allowed
+    /// to ignore the setting.
     pub cancellable: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 9]
```diff
@@ -3379,8 +3541,8 @@ pub struct WorkDoneProgressBegin {
 #[serde(rename_all = "camelCase")]
 pub struct WorkDoneProgressReport {
     /// Controls if a cancel button should show to allow the user to cancel the
-	/// long running operation. Clients that don't support cancellation are allowed
-	/// to ignore the setting.
+    /// long running operation. Clients that don't support cancellation are allowed
+    /// to ignore the setting.
     pub cancellable: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

**Comment by kjeremy** [comment]
`WorkDoneProgressDone` or whatever it's still called in this file is now `WorkDoneProgressEnd` so I would double check this and make sure progress is right.

**Comment by kjeremy** [comment]
Once accepted this PR will close #118

**Comment by kjeremy** [comment]
I don't see the new registration option stuff yet.

**Comment by kjeremy** [comment]
Also! I think you can remove the proposed feature around progress.

File: src/lib.rs [hunk 10]
```diff
@@ -777,6 +784,17 @@ pub struct InitializeParams {
     /// configured.
     #[serde(skip_serializing_if = "Option::is_none")]
     pub workspace_folders: Option<Vec<WorkspaceFolder>>,
+
+    /// Information about the client.
+    pub client_info: Option<ClientInfo>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 11]
```diff
@@ -777,6 +784,17 @@ pub struct InitializeParams {
...
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+pub struct ClientInfo {
+    /// The name of the client as defined by the client.
+    pub name: String,
+    /// The client's version as defined by the client.
+    pub version: Option<String>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 12]
```diff
@@ -1369,6 +1422,18 @@ pub struct ClientCapabilities {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub server_info: Option<ServerInfo>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+pub struct ServerInfo {
+    /// The name of the server as defined by the server.
+    pub name: String,
+    /// The servers's version as defined by the server.
+    pub version: Option<String>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 13]
```diff
@@ -1443,6 +1508,53 @@ pub struct SignatureHelpOptions {
...
+
+    /// `true` if signature help was already showing when it was triggered.
+    /// Retriggers occur when the signature help is already active and can be caused by actions such as
+    /// typing a trigger character, a cursor move, or document content changes.
+    pub is_retrigger: bool,
+
+    /// The currently active `SignatureHelp`.
+    /// The `activeSignatureHelp` has its `SignatureHelp.activeSignature` field updated based on
+    /// the user navigating through available signatures.
+    pub active_signature_help: Option<SignatureHelp>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 14]
```diff
@@ -3394,24 +3563,22 @@ pub struct WorkDoneProgressReport {
...
-#[cfg(feature = "proposed")]
 #[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
 #[serde(rename_all = "camelCase")]
-pub struct WorkDoneProgressDone {
+pub struct WorkDoneProgressEnd {
     /// Optional, more detailed associated progress message. Contains
     /// complementary information to the `title`.
     /// Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
     /// If unset, the previous progress message (if any) is still valid.
     pub message: Option<String>,
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(skip_serializing_if = "Option::is_none")]

File: src/lib.rs [hunk 15]
```diff
@@ -3581,6 +3597,13 @@ pub enum WorkDoneProgress {
     Done(WorkDoneProgressEnd),
 }
 
+/// A parameter literal used to pass a partial result token.
+#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
```
Comment (kjeremy) [inline src/lib.rs]: #[serde(rename_all = "camelCase")]

File: src/lib.rs [hunk 16]
```diff
@@ -1443,6 +1546,69 @@ pub struct SignatureHelpOptions {
...
+    /// Signature help was triggered by the cursor moving or by the document content changing.
+    ContentChange = 3,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SignatureHelpParams {
+    /// The signature help context. This is only available if the client specifies
+    /// to send this using the client capability  `textDocument.signatureHelp.contextSupport === true`
+    pub context: Option<SignatureHelpContext>,
```
Comment (kjeremy) [inline src/lib.rs]: `#[serde(skip_serializing_if = "Option::is_none")]`

File: src/lib.rs [hunk 17]
```diff
@@ -2944,6 +3207,18 @@ pub struct FormattingOptions {
     /// Signature for further properties.
     #[serde(flatten)]
     pub properties: HashMap<String, FormattingProperty>,
+
+    /// Trim trailing whitespaces on a line.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub trailing_whitespace: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: trim_trailing_whitespace

File: src/lib.rs [hunk 18]
```diff
@@ -2944,6 +3207,18 @@ pub struct FormattingOptions {
...
+
+    /// Insert a newline character at the end of the file if one does not exist.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub insert_final_new_line: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: I think this needs to be `insert_final_newline` to match the casing in the spec: `insertFinalNewline`

File: src/lib.rs [hunk 19]
```diff
@@ -2944,6 +3207,18 @@ pub struct FormattingOptions {
...
+
+    /// Trim all newlines after the final newline at the end of the file.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub trim_final_new_lines: Option<bool>,
```
Comment (kjeremy) [inline src/lib.rs]: Same here: `trim_final_newlines`

**Comment by kjeremy** [comment]
Besides the two fixes above what else is left on this? Looks like it's not marked draft anymore. Is it ready for @Marwes to review?

**Comment by Marwes** [comment]
Completely forgot about this, sorry! Released as 0.62

**Pull request merged**